### PR TITLE
Support deletes in StandaloneEncryptedFileSystem

### PR DESCRIPTION
**Pull request by ellisjoe**

### CODE REVIEW

**Comment by ellisjoe** [comment]
Fixes https://github.com/palantir/hadoop-crypto/issues/93

**Review by sandorw — APPROVED** [review]

File: hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java [hunk 1]
```diff
@@ -129,6 +131,14 @@ public boolean exists(Path path) throws IOException {
         return fs.exists(path);
     }
 
+    @Override
+    public boolean delete(Path path, boolean recursive) throws IOException {
+        // EncryptedFileSystem does not handle recursive deletes because it needs to rely on the KeyStorageStrategy.
+        // Here we know we are using the FileKeyStorageStrategy and can therefore delegate deletes to the FileSystem
+        // which will remove both the data and key material files.
+        return delegate.delete(path, recursive);
```
Comment (uschi2000) [inline hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java:139]: what if path is a file rather than a folder?

Comment (uschi2000) [reply hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java:139]: the key material won't get deleted, so this seems broken

File: hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java [hunk 2]
```diff
@@ -73,15 +73,17 @@ public boolean apply(FileStatus status) {
      * Key mapping to the public/private key algorithm (ex: RSA).
      */
     public static final String KEY_ALGORITHM_CONF = "fs.efs.key.algorithm";
+
     private String encryptedScheme;
+    private FileSystem delegate;
```
Comment (uschi2000) [inline hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java:78]: should explain what this is and how it's different from `this.fs`.

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java [hunk 3]
```diff
@@ -102,6 +103,49 @@ public void testReadWrite() throws IOException {
...
+        Path path1 = writeData(rootFolder);
+        Path path2 = writeData(rootFolder);
+
+        assertThat(efs.exists(path1)).isTrue();
+        assertThat(efs.exists(path2)).isTrue();
+
+        efs.delete(new Path(rootFolder.getAbsolutePath()), true);
+
+        assertThat(efs.exists(path1)).isFalse();
+        assertThat(efs.exists(path2)).isFalse();
```
Comment (uschi2000) [inline hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:135]: assertion about rootFolder missing

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java [hunk 4]
```diff
@@ -102,6 +103,49 @@ public void testReadWrite() throws IOException {
...
+    }
+
+    @Test
+    public void testRecursiveDelete() throws IOException {
+        File rootFolder = folder.newFolder();
+
+        Path path1 = writeData(rootFolder);
+        Path path2 = writeData(rootFolder);
+
+        assertThat(efs.exists(path1)).isTrue();
```
Comment (uschi2000) [inline hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:129]: move these to writeData?

Comment (uschi2000) [reply hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:129]: nah, actually fine here.

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java [hunk 5]
```diff
@@ -102,6 +103,49 @@ public void testReadWrite() throws IOException {
...
+        assertThat(efs.exists(path1)).isFalse();
+        assertThat(efs.exists(path2)).isFalse();
+    }
+
+    private Path writeData(File rootFolder) throws IOException {
+        // Write encrypted data
+        java.nio.file.Path filePath = Files.createTempFile(rootFolder.toPath(), "prefix", "suffix");
+        Path newPath = new Path(filePath.toAbsolutePath().toString());
+        OutputStream os = efs.create(path);
+        IOUtils.write("data", os);
```
Comment (uschi2000) [inline hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:143]: think you meant to write `big data` here.

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java [hunk 6]
```diff
@@ -102,6 +103,49 @@ public void testReadWrite() throws IOException {
         assertThat(rawFs.exists(new Path(path + FileKeyStorageStrategy.EXTENSION))).isTrue();
     }
 
+    @Test
+    public void testDelete() throws IOException {
+        File rootFolder = folder.newFolder();
+
+        Path path1 = writeData(rootFolder);
+        Path path2 = writeData(rootFolder);
```
Comment (uschi2000) [inline hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:111]: this seems a bit bogus: you're verifying that `delete` doesn't touch a totally unrelated path?

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java [hunk 7]
```diff
@@ -102,6 +103,49 @@ public void testReadWrite() throws IOException {
...
+
+        assertThat(efs.exists(path1)).isTrue();
+        assertThat(efs.exists(path2)).isTrue();
+
+        efs.delete(new Path(path1.toString()), false);
+
+        assertThat(efs.exists(path1)).isFalse();
```
Comment (uschi2000) [inline hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:118]: - should verify that you write to the same location again after deleting from it, I think.

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java (hunk 6 above)
Comment (ellisjoe) [reply hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:111]: removed in the follow up PR

File: hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java (hunk 1 above)
Comment (ellisjoe) [reply hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java:139]: fixed in the follow up

File: hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java (hunk 2 above)
Comment (ellisjoe) [reply hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystem.java:78]: added in follow up

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java (hunk 3 above)
Comment (ellisjoe) [reply hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:135]: done in follow up

File: hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java (hunk 7 above)
Comment (ellisjoe) [reply hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:118]: don't think that's relevant here? in my mind we've shown that `delete` puts the FileSystem back into a state where a file and it's key material at a given path do not exist. Our `create` tests should verify that we can create a new encrypted file from that state.

Comment (uschi2000) [reply hadoop-crypto/src/test/java/com/palantir/crypto2/hadoop/StandaloneEncryptedFileSystemTest.java:118]: ok, now that you're actually verifying that, that's fine :)

**Pull request merged**